### PR TITLE
Consolidate registry/wrapper argument validation

### PR DESCRIPTION
**Pull request by mattwthompson**
There's a lot of repeated code in the `Molecule` API that coaxes an argument that could be a `ToolkitWrapper` (or sometimes its class) into a `ToolkitRegistry`. The functionality here is identical across a few dozen uses except for some error messages that have subtly diverged over time. I took these all and wrapped them into a single function whose purpose should be clear from the signature:

```
def _ensure_toolkit_registry(
    toolkit_registry: Union[ToolkitRegistry, Type[ToolkitRegistry], ToolkitWrapper],
) -> ToolkitRegistry:
```

I don't think this breaks API. There might be some behavior changes resulting from a `ToolkitWrapper` being called instead of a `ToolkitRegistry` containing it, which arguably should have never happened.

This touches on some older discussions (https://github.com/openforcefield/openff-toolkit/issues/1066, #1067).

- [ ] Tag issue being addressed
- [ ] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/main/openff/toolkit/tests)
- [ ] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/main/docs), if applicable
- [ ] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [ ] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/main/docs/releasehistory.rst)

### CODE REVIEW

File: openff/toolkit/topology/molecule.py [hunk 1]
```diff
@@ -88,6 +91,9 @@
 #       Only support OEAroModel_MDL in RDKit version?
 
 
+TKR: TypeAlias = Union[ToolkitRegistry, ToolkitWrapper]
```
Comment (mattwthompson) [inline openff/toolkit/topology/molecule.py:94]: As I understand it, this is as simple as "hey, I'm re-using this type annotation a bunch of times, can I define it as variable-like thing so it's not so repetitive and verbose?"

https://peps.python.org/pep-0613/

File: openff/toolkit/topology/molecule.py [hunk 2]
```diff
@@ -1661,23 +1640,12 @@ def to_inchi(self, fixed_hydrogens=False, toolkit_registry=GLOBAL_TOOLKIT_REGIST
...
-            inchi = toolkit.to_inchi(self, fixed_hydrogens=fixed_hydrogens)
-        else:
-            raise InvalidToolkitRegistryError(
-                "Invalid toolkit_registry passed to to_inchi. Expected ToolkitRegistry or ToolkitWrapper. "
-                f"Got {type(toolkit_registry)}"
-            )
+        toolkit_registry = _ensure_toolkit_registry(toolkit_registry)
 
-        return inchi
+        return toolkit_registry.call("to_inchi", self, fixed_hydrogens=fixed_hydrogens)
```
Comment (mattwthompson) [inline openff/toolkit/topology/molecule.py:1645]: This pattern could be further golfed down to a one-liner


```
return _ensure_toolkit_registry(toolkit_registry)..call("method_name", ...)
```
but I think that's too far

Comment (j-wags) [reply openff/toolkit/topology/molecule.py:1645]: Haha yeah, it's already golfy enough :-P

**Review by j-wags — APPROVED** [review]
Thanks @mattwthompson - These PRs are like an advent calendar - I love getting negative lines of code in my stocking :-P 

This _technicalllly_ changes the exception type of some `Molecule` API calls by having all of these route through `ToolkitRegistry` instead of having some route through `ToolkitWrapper` (depending on the input type). For example, calling `RDKitToolkitWrapper.generate_conformers` could raise a `ConformerGenerationError`. But `ToolkitRegistry([RDKitToolkitWrapper()]).call('generate_conformers'...)` would raise [a "summary" `ValueError`](https://github.com/openforcefield/openff-toolkit/blob/efe91cd54b8bae0e5046b1366bb6b11a5447464d/openff/toolkit/utils/toolkit_registry.py#L359-L370) that describes the `ConformerGenerationError` raised by the toolkitwrappers. 

Anyway, I don't care that much, let's just be a responsible upstream by mentioning this in the releasehistory and make the next release be a minor version bump.

**Comment by mattwthompson** [comment]
Do you care much about `FrozenMolecule.from_file`? I should have mentioned this before; that's the only method including this logic that I didn't touch. The special cases between file formats and wrapped toolkits looked pretty fragile to me.

**Review by j-wags — CHANGES_REQUESTED** [review]
Oh, I'd forgotten about all that. Let me think a bit more and test some stuff out.

**Review by j-wags — COMMENTED** [review]
Ok, so I think that anywhere we used to specify `raise_exception_types=[]` (namely `Molecule.assign_partial_charges` and `generate_conformers`) should be reverted to their original form, and also `Molecule.from_file` should remain in its original form. Those methods have specific behaviors when using toolkits/throwing errors that I think are useful enough to keep around. But all the other changes should be fine to stay.

**Comment by mattwthompson** [comment]
Which are the existing behaviors you'd like to preserve?

**Comment by mattwthompson** [comment]
I've run slow tests locally and everything is passing there as well. (With my luck something will fail in CI, of course.)

**Review by j-wags — COMMENTED** [review]
Two hours ago I came by to give this a quick approval but I saw that some good questions were added about the architecture of the Molecule API. I started typing up a big response but after a while I realized that architecture discussion would be lost and architecture change is out of scope of this PR, which is just intended to simplify some reused code. But this PR doesn't really accomplish that because it's changing more than 300 lines of in order to remove maybe 40 lines of source, so I don't think it really succeeds there. It also adds type annotations, which I like, but I don't like the changes to get_tagged_smarts_connectivity, which seem off topic. I'm gonna close this PR because I think the cost/benefit here has gotten a bit too high, but I'd be happy to consider these changes as separate PRs in the future.

**Comment by mattwthompson** [comment]
Nothing changed with `get_tagged_smarts_connectivity` but I can revert those if you wish?

**Pull request closed without merging**